execution_engine/src/shared/gas.rs

```diff
@@ -1,36 +1,38 @@
 use std::{fmt, iter::Sum};
 
-use num::Zero;
+use num::{CheckedMul, Integer, Zero};
+use num_derive::{Num, NumOps, One, Zero};
+use num_traits::{CheckedAdd, CheckedSub};
 
 use casper_types::U512;
 
 use crate::shared::motes::Motes;
 
-#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
-pub struct Gas(U512);
+const GAS_MAX_AS_U512: U512 = U512([u64::MAX, 0, 0, 0, 0, 0, 0, 0]);
+
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Num, Zero, One, NumOps)]
```

Comment:
Why would we ever want to do a regular add/sub?  Can we get rid of `NumOps` here?

Reply:
That would be great to prevent using unchecked arithmetic, although opting out of NumOps would also require opting out of Zero (and others) as this one requires Add to be implemented etc. I don't know - I think we should either go all-in into num traits or opt-out of those completely and implement only checked ops using methods and no traits. What do you think?